join-domain/windows/init.sls

```
@@ -1,46 +1,9 @@
-        {
-          "changed=no comment=`"System is joined already to the correct domain
-            [$domain].`" domain=$domain";
-        }
-        else
-        {
-          throw "System is joined to another domain [$domain]. To join a
-            different domain, first remove it from the current domain."
-        }
-      }
-      else
-      {
-        $AesObject = New-Object System.Security.Cryptography.AesCryptoServiceProvider;
-        $AesObject.IV = New-Object Byte[]($AesObject.IV.Length);
-        $AesObject.Key = [System.Convert]::FromBase64String("{{ join_domain.key }}");
-        $EncryptedStringBytes = [System.Convert]::FromBase64String(
-          "{{ join_domain.encrypted_password }}" );
-        $cred = New-Object -TypeName System.Management.Automation.PSCredential
-          -ArgumentList {{ join_domain.username }}, (ConvertTo-SecureString
-          -String "$([System.Text.UnicodeEncoding]::Unicode.GetString(
-          ($AesObject.CreateDecryptor()).TransformFinalBlock($EncryptedStringBytes,
-          0, $EncryptedStringBytes.Length)))"
-          -AsPlainText -Force);
-    {%- if join_domain.oupath -%}
-        Add-Computer -DomainName {{ join_domain.dns_name }} -Credential $cred
-          -OUPath "{{ join_domain.oupath }}"
-          -Options JoinWithNewName,AccountCreate -Force -ErrorAction Stop;
-    {%- else -%}
-        Add-Computer -DomainName {{ join_domain.dns_name }} -Credential $cred
-          -Options JoinWithNewName,AccountCreate -Force -ErrorAction Stop;
-    {%- endif -%}
-        "changed=yes comment=`"Joined system to the domain.`"
-        domain={{ join_domain.dns_name }}"
-      }'
+  cmd.script:
+    - name: salt://{{ tpldir }}/files/JoinDomain.ps1
+    - args: -DomainName "{{ join_domain.dns_name }}" -TargetOU "{{ join_domain.oupath }}" -Key "{{ join_domain.key }}" -EncryptedPassword "{{ join_domain.encrypted_password }}" -UserName "{{ join_domain.username }}"
```

Maybe add `-ErrorAction Stop` to the args and re-run the tests... Curious if we have non-fatal errors that the default stop action is masking.

---

Ok.  I'll check.

---

Adding `-ErrorAction Stop` to the args didn't produce any errors.  The tests all completed successfully.  Unless there are any objections, I will keep the change and push an update.

---

Sounds good!